src/bootstrap/src/core/build_steps/compile.rs

```diff
@@ -1957,7 +1954,12 @@ pub fn run_cargo(
             let file_stem = parts.next().unwrap().to_owned();
             let extension = parts.next().unwrap().to_owned();
 
-            toplevel.push((file_stem, extension, expected_len));
+            if extension == "so" || extension == "dylib" {
+                // FIXME workaround for the fact that cargo doesn't understand `-Zsplit-metadata`
+                toplevel.push((file_stem.clone(), "rmeta".to_owned(), None));
+            }
```

Comment:
This together with the rustc wrapper passing `--emit metadata` is a big hack working around the fact that cargo doesn't support `-Zsplit-metadata` and as such doesn't cause a `.rmeta` file to be emitted and also doesn't track it in the compiler-artifact json message. This also causes issues with recompilations as the loop below may copy outdated rmeta files into the sysroot, which would then result in compiler errors.